src/api/app/controllers/webui/project_controller.rb

```diff
@@ -606,52 +558,86 @@ def unlock
 
   private
 
-  def fill_status_cache
-    @repohash = {}
+  def monitor_buildresult
+    @legend = Buildresult::STATUS_DESCRIPTION
+
+    @name_filter = params[:pkgname]
+    @lastbuild_switch = params[:lastbuild]
+    if params[:defaults]
+      defaults = (begin
+                    Integer(params[:defaults])
+                  rescue ArgumentError
+                    1
+                  end) > 0
+    else
+      defaults = true
+    end
+    params['expansionerror'] = 1 if params['unresolvable']
+    monitor_set_filter(defaults)
+
+    find_opt = { project: @project, view: 'status', code: @status_filter,
+                 arch: @arch_filter, repository: @repo_filter }
+    find_opt[:lastbuild] = 1 if @lastbuild_switch.present?
+
+    buildresult = Buildresult.find_hashed(find_opt)
+    if buildresult.empty?
+      flash[:warning] = "No build results for project '#{@project}'"
+      redirect_to action: :show, project: params[:project]
+      return
+    end
+
+    return unless buildresult.key?('result')
+    buildresult
+  end
+
+  def monitor_parse_buildresult(buildresult)
+    @packagenames = Set.new
```

Comment:
why a set? before we were not taking care of repetitions and sorting is less efficient now as you need to convert to an array....

Reply:
we were taking care of repetitions. This set replaces

```
    @packagenames = @packagenames.flatten.uniq.sort!

    ## Filter for PackageNames ####
    @packagenames.select! { |name| filter_matches?(name, @name_filter) } if @name_filter.present?